[export] refactor _process_dynamic_shapes

pianpwk opened the pull request:
Sorryyyyy for another refactor. This splits `_process_dynamic_shapes` into 3 parts:
1. `_combine_args` - mostly the same thing
2. `_check_dynamic_shapes`, which is responsible for raising 99% of UserErrors if the dynamic shapes spec is invalid (minus 1 UserError with DerivedDims)
3.  `_process_dynamic_shapes`, which for now, is the same thing, minus the stuff in 2.

This refactor is helpful for incoming automatic dynamic shapes work, because, we're switching to `assume_static_by_default=False`, which is what `_dynamo.export` currently does. This means any unspecified dims are allocated a symbol, in contrast to export today which keeps unspecified dims static. Historically this has been desirable - export users don't want too much dynamism. So we want to change how the spec is translated into constraints.
 
This means when we switch over to automatic dynamic shapes, we want to plug in something in between steps 2. and 3. which patches up the spec for `assume_static_by_default=False`, filling in static shapes for any unspecified dims, and potentially clearing out the auto-dynamic dims (since they're no-ops). We would do this in-between 2. and 3. to keep `_process_dynamic_shapes` semantically the same, since it's used with `_dynamo.export`.

We could do this without a refactor, plugging in this transform before `_process_dynamic_shapes`, but since that function's responsible for both spec checking + constraint production, moving spec checking to before we transform the specs helps guarantee we're raising errors on what the user's specified, and not an internal export bug.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

avikchaudhuri commented on torch/_export/non_strict_utils.py:
```diff
@@ -120,10 +125,11 @@ def make_fake_inputs(
     )
-    constraints = constraints or []
+    _check_dynamic_shapes(combined_args, dynamic_shapes)
+    constraints = _process_dynamic_shapes(combined_args, dynamic_shapes) or []
```
Is there any material difference between `None` and `[]` here? I suspect not, so maybe take this opportunity to return `[]` everywhere `None` was returned and kill this `or`.

avikchaudhuri commented on torch/export/dynamic_shapes.py:
```diff
@@ -796,17 +796,128 @@ def find_shape(path, t):
-    _is_torch_jit_trace=False,
+    combined_args: Dict[str, Any],
+    dynamic_shapes: Union[Dict[str, Any], Tuple[Any], List[Any], None],
 ) -> Optional[List[Constraint]]:
```
Change this type to `List[Constraint]` if you believe returning `None` is equivalent to returning `[]` here.

avikchaudhuri reviewed (approved):
Looks great!

avikchaudhuri commented:
I like the split into smaller functions and buy this is useful for incrementally changing behavior, but frankly I didn't quite grok the part of the summary that talks about `assume_static_by_default=False`:

> because, we're switching to assume_static_by_default=False, which is what _dynamo.export currently does. This means any unspecified dims are allocated a symbol, in contrast to export today which keeps unspecified dims static. Historically this has been desirable - export users don't want too much dynamism. So we want to change how the spec is translated into constraints.

> This means when we switch over to automatic dynamic shapes, we want to plug in something in between steps 2. and 3. which patches up the spec for assume_static_by_default=False, filling in static shapes for any unspecified dims, and potentially clearing out the auto-dynamic dims (since they're no-ops). 

Can you explain again? Are you taking about choices in DimDynamic cases here? 

We have potentially two candidate changes: (1) we take dims marked dynamic but otherwise unspecified in terms of constraints they satisfy, and we just accumulate them (2) with multiple sample inputs, potentially upgrade a static to a dynamic. I believe we're mostly going to focus on (1) for now. How does `assume_static_by_default=False` influence this?

pianpwk committed:
[export] refactor _process_dynamic_shapes (#133391)

Summary:
Sorryyyyy for another refactor. This splits `_process_dynamic_shapes` into 3 parts:
1. `_combine_args` - mostly the same thing
2. `_check_dynamic_shapes`, which is responsible for raising 99% of UserErrors if the dynamic shapes spec is invalid (minus 1 UserError with DerivedDims)
3.  `_process_dynamic_shapes`, which for now, is the same thing, minus the stuff in 2.

This refactor is helpful for incoming automatic dynamic shapes work, because, we're switching to `assume_static_by_default=False`, which is what `_dynamo.export` currently does. This means any unspecified dims are allocated a symbol, in contrast to export today which keeps unspecified dims static. Historically this has been desirable - export users don't want too much dynamism. So we want to change how the spec is translated into constraints.
 
This means when we switch over to automatic dynamic shapes, we want to plug in something in between steps 2. and 3. which patches up the spec for `assume_static_by_default=False`, filling in static shapes for any unspecified dims, and potentially clearing out the auto-dynamic dims (since they're no-ops). We would do this in-between 2. and 3. to keep `_process_dynamic_shapes` semantically the same, since it's used with `_dynamo.export`.

We could do this without a refactor, plugging in this transform before `_process_dynamic_shapes`, but since that function's responsible for both spec checking + constraint production, moving spec checking to before we transform the specs helps guarantee we're raising errors on what the user's specified, and not an internal export bug.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames


Reviewed By: avikchaudhuri

Differential Revision: D61255027

Pulled By: pianpwk

pianpwk commented:
> I like the split into smaller functions and buy this is useful for incrementally changing behavior, but frankly I didn't quite grok the part of the summary that talks about `assume_static_by_default=False`:
> 
> > because, we're switching to assume_static_by_default=False, which is what _dynamo.export currently does. This means any unspecified dims are allocated a symbol, in contrast to export today which keeps unspecified dims static. Historically this has been desirable - export users don't want too much dynamism. So we want to change how the spec is translated into constraints.
> 
> > This means when we switch over to automatic dynamic shapes, we want to plug in something in between steps 2. and 3. which patches up the spec for assume_static_by_default=False, filling in static shapes for any unspecified dims, and potentially clearing out the auto-dynamic dims (since they're no-ops).
> 
> Can you explain again? Are you taking about choices in DimDynamic cases here?
> 
> We have potentially two candidate changes: (1) we take dims marked dynamic but otherwise unspecified in terms of constraints they satisfy, and we just accumulate them (2) with multiple sample inputs, potentially upgrade a static to a dynamic. I believe we're mostly going to focus on (1) for now. How does `assume_static_by_default=False` influence this?

Sorry it's not so clear, this doesn't relate to (2) yet, I'm still not sure how to do that with export.

This has to do with dynamic_shapes specs like `{"x": None}`. Traditionally for export.export(), this means x's shapes are static. But switching over to `assume_static_by_default=False`, this treats x as entirely dynamic (symbols allocated for all dims). We could change this, modifying _process_dynamic_shapes to treat None as static, but `_dynamo.export()` is relying on this behavior, already using `assume_static_by_default=False`. So I'm relying on this spec transformation to convert export shapes specs into something compatible with dynamic by default.

The transformation in between 2 & 3 I'm talking about is mainly to transform Nones into explicitly static dims `{"x": None} -> {"x": 3}`, for example. Also we would transform Trues (auto-dynamic dims) into Nones (since None -> dynamic by default). This is kinda essentially inverting the spec going into _process_dynamic_shapes, so it knows to leave auto-dynamic dims alone, and create static constraints for None/unspecified dims from export.export().

I guess the specs are semantically very similar to `_dynamo.export()` now, just with this extra transformation to treat unspecified/None dims as static for export.export(). What the specs are is essentially (from my understanding):

For export.export():
- True: auto-dynamic, do whatever, don't raise constraint violations
- None: static
- Dims/DerivedDims: very strictly assert on the min/max range for this symbol, and require a specification for all dims governed by this symbol (i.e. relations, equality, linear relations).

For _dynamo.export():
- None: auto-dynamic
- Dims/DerivedDims: also a strict assertion

avikchaudhuri commented:
OK, this is clearer. Are you sure `None` is even reachable in `_dynamo.export()` though? I'd imagine the only client is `export.export` which makes things super explicit with `assume_static_by_default=True`, and we killed all paths into `_dynamo.export()` that don't go through `process_dynamic_shapes`.

pianpwk commented:
> OK, this is clearer. Are you sure `None` is even reachable in `_dynamo.export()` though? I'd imagine the only client is `export.export` which makes things super explicit with `assume_static_by_default=True`, and we killed all paths into `_dynamo.export()` that don't go through `process_dynamic_shapes`.

Yeah, looking at [test/dynamo/test_export.py](https://github.com/pytorch/pytorch/blob/main/test/dynamo/test_export.py) it does seem possible to specify None for some inputs. Also for unspecified, the handling is differently.
